### PR TITLE
Count physical cores properly on linux

### PR DESCRIPTION
**Pull request by pedrocr**
A possible way to do #30 with simpler code

### CODE REVIEW

**Comment by seanmonstar** [comment]
I find this much easier to follow than #30, thanks!

If I understand correctly, won't this have a different issue where it's counting virtual cores as physical? It seems like a machine with 1 physical core and 2 virtual will report as 2 physical cores, since it will store `(0, 0)` and `(0, 1)`, and then return `map.len()`.

Is the right thing to do to just only ever count `physical id`? Or is there cases where that key is not present?

**Comment by pedrocr** [comment]
That machine would work correctly. You'd have (0,0) and (0,0) for both processors as they are threads on a sing core, so both physical id and core id are the same. If you only count physical id on machines like most dual core processors you'd get 1 instead of 2, because you have a single physical id with two core id's. Doing the HashSet like in #32 is probably also a good idea.

**Comment by seanmonstar** [comment]
I'm not following what you mean. I don't have a Linux machine available that has multiple physical cores, so I can't test myself (just 1 virtual core). But it seems like cpuinfo could contain info like this:

```
physical id : 0
siblings    : 2
core id     : 0

// ...

physical id : 0
siblings    : 2
core id     : 1
```

And in this case, while only have 1 physical core, the map will have an entry for `(0, 0)` and `(0, 1)`, resulting in a count of 2 physical cpus.

If that's not the case, is there a resource explaining it?

----

Yes, a `HashSet` feels more correct (though a `HashSet` is just `HashMap<T, ()>`). I hope to merge that, but wanted to make sure it didn't cause conflicts with this work first.

**Comment by pedrocr** [comment]
That case is correct. You have a single physical cpu package, with two physical cores inside with a single thread each. So you do have 2 physical cores.

**Comment by seanmonstar** [comment]
So cpuinfo will never report about virtual cores in this manner? Sorry for my confusion here, it's not immediately obvious.

**Comment by pedrocr** [comment]
Here's an example from my machine. I have a laptop with a dual-core cpu and each of those cores has hyperthreading so has 2 threads:

 ```sh
$ grep -E 'physical id|core id|siblings' /proc/cpuinfo 
physical id	: 0
siblings	: 4
core id		: 0
physical id	: 0
siblings	: 4
core id		: 1
physical id	: 0
siblings	: 4
core id		: 0
physical id	: 0
siblings	: 4
core id		: 1
```

So I have four threads in total and thus four entries in /proc/cpuinfo. All the CPUs share the same physical_id. Every two threads share the same core id. All the threads report 4 siblings. The previous code worked fine for this situation but not when you had two different physical id's. 

Basically (physical_id, core_id) identifies a physical core. All the cores that share a physical_id are sharing the same package (i.e., in a dual CPU motherboard are attached in the same slot). Hyperthreading is identified by repeating (physical_id, core_id) pairs and the siblings count also includes hyperthreading.

**Comment by seanmonstar** [comment]
Published in `1.2.1`.
